app/controllers/concerns/shopify_app/jwt_authenticated.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module ShopifyApp
+  module JwtAuthenticated
+    extend ActiveSupport::Concern
+
+    included do
+      include ShopifyApp::Localization
+      include ShopifyApp::CsrfProtection
+      include ShopifyApp::EmbeddedApp
+      before_action :login_again_if_different_user_or_shop
+      around_action :activate_shopify_session
+    end
+
+    ACCESS_TOKEN_REQUIRED_HEADER = 'X-Shopify-API-Request-Failure-Unauthorized'
+
+    def activate_shopify_session
+      if user_expected? && user.blank?
+        return signal_access_token_required
+      end
+
+      # TODO: Temporary until we have redirect-less Offline Access Token support
+      return redirect_to_login if current_shopify_session.blank?
+
+      begin
+        ShopifyAPI::Base.activate_session(current_shopify_session)
+        yield
+      ensure
+        ShopifyAPI::Base.clear_session
+      end
+    end
+
+    def current_shopify_session
+      @current_shopify_session ||=
+        begin
+          shopify_session || user || shop
+        end
+    end
+
+    def shopify_session
+      # TODO: Configure a "true" Shopify session using the `sid` value from JWT
```

Comment:
I'll try doing this in my POC within the old login protection and see if it works